### PR TITLE
 Add parse method to S3FileStorageOptions so that we can config it as connectionString

### PR DESCRIPTION
**Pull request by edwardmeng**

### CODE REVIEW

**Comment by niemyjski** [comment]
I'm thinking that we have a parse on the options but we could have a utility method like this that can be used by the parsing for both queues and storage

```c#
    public class ConnectionStringHelper {
        public static AWSConnectionInfo GetAWSConnectionInfo(string connectionString) {
            if (String.IsNullOrEmpty(connectionString))
                return new AWSConnectionInfo();

            var settings = ParseConnectionString(connectionString);
            
            var connectionInfo = new AWSConnectionInfo();
            if (settings.ContainsKey("AccessKey"))
                connectionInfo.AccessKey = settings["AccessKey"];
            if (settings.ContainsKey("SecretKey"))
                connectionInfo.SecretKey = settings["SecretKey"];
            if (settings.ContainsKey("Bucket"))
                connectionInfo.Bucket = settings["Bucket"];
            if (settings.ContainsKey("Region"))
                connectionInfo.Region = settings["Region"];
            
            return connectionInfo;
        }

        public static Dictionary<string, string> ParseConnectionString(string connectionString) {
            return connectionString.Split(';')
                .Where(kvp => kvp.Contains('='))
                .Select(kvp => kvp.Split(new[] { '=' }, 2))
                .ToDictionary(kvp => kvp[0].Trim(), kvp => kvp[1].Trim(), StringComparer.InvariantCultureIgnoreCase);
        }
    }

    public class AWSConnectionInfo {
        public string AccessKey { get; set; }
        public string SecretKey { get; set; }
        public string Bucket { get; set; }
        public string Region { get; set; }

        public RegionEndpoint GetRegionEndpoint() {
            return RegionEndpoint.GetBySystemName(String.IsNullOrEmpty(Region) ? "us-east-1" : Region);
        }

        public AWSCredentials GetCredentials() {
            if (String.IsNullOrEmpty(AccessKey)
                || String.IsNullOrEmpty(SecretKey))
                return FallbackCredentialsFactory.GetCredentials();

            return new BasicAWSCredentials(AccessKey, SecretKey);
        }
    }
```

**Comment by niemyjski** [comment]
Also, I'm curious if we are creating more work for ourselves by allowing multiple variations of each key / connection strings (unless there are connection string formats out there like this already)

**Comment by edwardmeng** [comment]
I have completed the changes, please take review.

File: src/Foundatio.AWS/Metrics/CloudWatchMetricsClientOptions.cs [hunk 1]
```diff
@@ -1,14 +1,18 @@
 using System;
-using System.Collections.Generic;
-using Amazon;
-using Amazon.CloudWatch.Model;
-using Amazon.Runtime;
 
 namespace Foundatio.Metrics {
     public class CloudWatchMetricsClientOptions : MetricsClientOptionsBase {
-        public AWSCredentials Credentials { get; set; }
```
Comment (ejsmith) [inline src/Foundatio.AWS/Metrics/CloudWatchMetricsClientOptions.cs]: I think you should be able to set the AccessKey, SecretKey, and Region manually on the options object in addition to being able to set the connection string.

File: src/Foundatio.AWS/AmazonConnection.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+using System.Linq;
+using Amazon;
+using Amazon.Runtime;
+
+namespace Foundatio {
+    internal class AmazonConnection {
+        public AWSCredentials Credentials { get; set; }
+        public RegionEndpoint Region { get; set; }
+
+        protected static T Parse<T>(string connectionString, Func<string, string, T, bool> itemParser = null) where T : AmazonConnection, new() {
```
Comment (ejsmith) [inline src/Foundatio.AWS/AmazonConnection.cs]: I like it. Nice job.

File: src/Foundatio.AWS/Metrics/CloudWatchMetricsClientOptions.cs (hunk 1 above)
Comment (edwardmeng) [reply src/Foundatio.AWS/Metrics/CloudWatchMetricsClientOptions.cs]: I think we should keep the same design pattern for AWS, Aliyun and Azure. The AccessKey, SecretKey, and Region are configured with ConnectionString in Aliyun and Azure, but only AWS is not.

**Comment by edwardmeng** [comment]
I have added connection string builders for every module which should be used for setting the connection options dynamically.

**Comment by edwardmeng** [comment]
I have completed the changes, please take review again.

**Comment by ejsmith** [comment]
The tests are failing because the config is failing to pickup the settings from environment variables. On the builder server we have AWS:ACCESS_KEY_ID and AWS:SECRET_ACCESS_KEY configured. It looks like the tests are picking those variables up and using them, but something is broken.

**Comment by edwardmeng** [comment]
I have added some unit test methods for the connection string builders. The failed tests caused by incorrect connection string syntax in the unit tests.
Please take review again.

**Pull request merged**